File: app/src/main/java/com/example/newsapp/viewmodel/NewsViewModelFactory.java
package com.example.newsapp.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.example.newsapp.repository.NewsRepository;

public class NewsViewModelFactory implements ViewModelProvider.Factory {
    private NewsRepository newsRepository;

    public NewsViewModelFactory(NewsRepository newsRepository) {
        this.newsRepository = newsRepository;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> aClass) {
        return (T) new NewsViewModel(newsRepository);
    }
}

File: app/src/main/java/com/example/newsapp/viewmodel/NewsViewModelFactory.kt
package com.example.newsapp.viewmodel

import com.example.newsapp.repository.NewsRepositoryImpl
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

class NewsViewModelFactory(private val newsRepository: NewsRepositoryImpl) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(NewsViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return NewsViewModel(newsRepository) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}